### PR TITLE
Linear equivalence classes

### PR DESCRIPTION
**Pull request by valentin-pinkau**
This PR computes the equivalence classes and returns the offsets within the linear equivalence classes.

### CODE REVIEW

File: networkit/cpp/components/ConnectedComponents.cpp [hunk 1]
```diff
@@ -62,14 +62,45 @@ cc_result ConnectedComponents::get_raw_partition(const Graph & G) {
         }
     });
 
-    auto mapping_sizes = new node[n_components];
+    // add one to reserve memory for additional last offset (size)
+    auto mapping_sizes = new node[n_components + 1];
```
Comment (georgwiese) [inline networkit/cpp/components/ConnectedComponents.cpp:66]: Why is the type `node`, not `int`? I'm guessing it's equivalent, but this is confusion, no?

File: networkit/cpp/components/ConnectedComponents.cpp [hunk 2]
```diff
@@ -62,14 +62,45 @@ cc_result ConnectedComponents::get_raw_partition(const Graph & G) {
...
+
     for(node n = 0; n < max_id; ++n) {
         ++mapping_sizes[mapping_array[n]];
     }
+    // set size of component 0 to 0
+    mapping_sizes[0] = 0;
+
+    // sizes:   1, 2, 5, 6, 3
+    // offsets: 0, 1, 3, 8, 14
+    node sum = 0;
```
Comment (georgwiese) [inline networkit/cpp/components/ConnectedComponents.cpp:77]: Again, why is `sum` of type `node`?

File: networkit/cpp/components/ConnectedComponents.cpp [hunk 3]
```diff
@@ -62,14 +62,45 @@ cc_result ConnectedComponents::get_raw_partition(const Graph & G) {
...
+    auto offsets = mapping_sizes;
+    for(node component_id = 0; component_id < n_components; ++component_id) {
+        auto tmp = sum;
+        sum += mapping_sizes[component_id];
+        offsets[component_id] = tmp;
+    }
+    auto equivalence_classes = new node[max_id];
+    // reinterpret offsets as index and create equivalence classes
+    // within a class nodes are sorted
+    auto in_class_index = offsets;
```
Comment (georgwiese) [inline networkit/cpp/components/ConnectedComponents.cpp:88]: Why don't we copy here? For memory efficiency?

**Review by georgwiese — APPROVED** [review]
LGTM!

File: networkit/cpp/components/ConnectedComponents.cpp (hunk 1 above)
Comment (valentin-pinkau) [reply networkit/cpp/components/ConnectedComponents.cpp:66]: The `node` type is `uint_64`. As there exists the possibility that all nodes are in the same component, I would like to use the same type as for the node ids which is `node`.

File: networkit/cpp/components/ConnectedComponents.cpp (hunk 2 above)
Comment (valentin-pinkau) [reply networkit/cpp/components/ConnectedComponents.cpp:77]: Same reason as above

File: networkit/cpp/components/ConnectedComponents.cpp (hunk 3 above)
Comment (valentin-pinkau) [reply networkit/cpp/components/ConnectedComponents.cpp:88]: I decided to do so for memory efficiency.

**Pull request merged**